sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala

```diff
@@ -56,9 +57,14 @@ class FailureSafeParser[IN](
     }
   }
 
+  private val skipParsing = optimizeEmptySchema && schema.isEmpty
   def parse(input: IN): Iterator[InternalRow] = {
     try {
-      rawParser.apply(input).toIterator.map(row => toResultRow(Some(row), () => null))
+     if (skipParsing) {
+       Iterator.single(InternalRow.empty)
+     } else {
+       rawParser.apply(input).toIterator.map(row => toResultRow(Some(row), () => null))
```

Comment:
If there are broken records the parser can't parse, this skipping won't detect them?

Reply:
Could you add a test case for counting both CSV and JSON source when the files having broken records? Any behavior change after this PR?

Reply:
> ... when the files having broken records?

Syntactically broken or semantically (wrong types for example)?

> Any behavior change after this PR?

We have many tests in `CSVSuite` and `JSONSuite` for broken records. I have found behavior change in only one case: https://github.com/apache/spark/pull/21909/files#diff-fde14032b0e6ef8086461edf79a27c5dL2227 . This is due to `Jackson` parser touches a few first bytes in the input stream even if it is not called. `Jackson` checks encoding eagerly.

Reply:
both? 

If we introduce a behavior change, we need to document it in the migration guide and add a conf. Users can do the conf to revert back to the previous behaviors.

Reply:
I added the tests